src/enclave/tls_endpoint.h

```diff
@@ -568,9 +561,136 @@ namespace enclave
       return reinterpret_cast<TLSEndpoint*>(ctx)->handle_recv(buf, len);
     }
 
+    // These callbacks below are complex, using the callbacks above and
+    // manipulating OpenSSL's BIO objects accordingly. This is just so we can
+    // emulate MbedTLS.
+    // Once we get rid of MbedTLS we can move the callbacks above to handle BIOs
+    // directly and hopefully remove the complexity below.
+#ifndef TLS_PROVIDER_IS_MBEDTLS
+    static long send_callback_openssl(
+      BIO* b,
+      int oper,
+      const char* argp,
+      size_t len,
+      int argi,
+      long argl,
+      int ret,
+      size_t* processed)
+    {
+      // Unused arguments
+      (void)argi;
+      (void)argl;
+      (void)argp;
+
+      if (ret && len > 0 && oper == (BIO_CB_WRITE | BIO_CB_RETURN))
+      {
+        // Flush BIO so the "pipe doesn't clog", but we don't use the
+        // data here, because 'argp' already has it.
+        BIO_flush(b);
+        size_t pending = BIO_pending(b);
+        if (pending)
+          BIO_reset(b);
+
+        // Pipe object
+        void* ctx = (BIO_get_callback_arg(b));
+        int put = send_callback(ctx, (const uint8_t*)argp, len);
+
+        // WANTS_WRITE
+        if (put == TLS_WRITING)
+        {
+          LOG_TRACE_FMT("TLS Endpoint::send_cb() : WANTS_WRITE");
+          *processed = 0;
+          return -1;
```

Comment:
What is the -1 here? Is that equivalent to `return WANTS_WRITE;`?

Reply:
Not always, but in this case, yes. `-1` just means `error`, but in this case the error can only be `WANTS_WRITE`, thus the log line.

This is one of the complications of using callbacks in OpenSSL. The callback has to be super careful to return the expected value as well as update the `processed` value or it breaks the rest of the read/write logic in `BIO_read_ex` and `BIO_write_ex`. This is why there's a complex manipulation of return values on both send and receive callbacks.